src/ray/raylet/test/pubsub_client_test.cc

```diff
@@ -0,0 +1,381 @@
+  // Callback is not invoked.
+  ASSERT_EQ(object_subscribed_.count(unrelated_object_id), 0);
+  ASSERT_TRUE(owner_client->ReplyObjectEviction(Status::NotFound("")));
+  ASSERT_EQ(object_failed_to_subscribe_.count(object_id), 1);
+
+  // The second long polling should fail.
+  object_failed_to_subscribe_.clear();
+  ASSERT_TRUE(owner_client->ReplyLongPolling(objects_batched, Status::NotFound("")));
+  // Since the failure callback is already invoked, it shouldn't be invoked again.
+  ASSERT_EQ(object_failed_to_subscribe_.count(object_id), 0);
+}
+
+TEST_F(PubsubClientTest, TestSubscriptionSucceedLongPollingFailed) {
+  auto subscription_callback = [this](const ObjectID &object_id) {
+    object_subscribed_.emplace(object_id);
+  };
+  auto failure_callback = [this](const ObjectID &object_id) {
+    object_failed_to_subscribe_.emplace(object_id);
+  };
+
+  const auto owner_addr = GenerateOwnerAddress();
+  const auto object_id = ObjectID::FromRandom();
+  pubsub_client_->SubcribeObject(owner_addr, object_id, subscription_callback,
+                                 failure_callback);
+  ASSERT_TRUE(owner_client->ReplyObjectEviction());
+
+  // Long polling failed.
+  std::vector<ObjectID> objects_batched;
+  ASSERT_TRUE(owner_client->ReplyLongPolling(objects_batched, Status::NotFound("")));
+  // Callback is not invoked.
+  ASSERT_EQ(object_subscribed_.count(object_id), 0);
+  // Failure callback is invoked.
+  ASSERT_EQ(object_failed_to_subscribe_.count(object_id), 1);
+  // Since the long polling is failed due to the owner failure, we shouldn't have any
+  // outstanding long polling request.
+  ASSERT_EQ(owner_client->GetNumberOfInFlightLongPollingRequests(), 0);
+}
+
+TEST_F(PubsubClientTest, TestUnsubscribeInSubscriptionCallback) {
+  const auto owner_addr = GenerateOwnerAddress();
+  const auto object_id = ObjectID::FromRandom();
+  // Test unsubscription call inside the subscription callback doesn't break raylet.
+  auto subscription_callback = [this, owner_addr](const ObjectID &object_id) {
+    pubsub_client_->UnsubscribeObject(owner_addr, object_id);
+    object_subscribed_.emplace(object_id);
+  };
+  auto failure_callback = [](const ObjectID &object_id) {
+    // This shouldn't be invoked in this test.
+    ASSERT_TRUE(false);
+  };
+
+  pubsub_client_->SubcribeObject(owner_addr, object_id, subscription_callback,
+                                 failure_callback);
+  ASSERT_TRUE(owner_client->ReplyObjectEviction());
+  std::vector<ObjectID> objects_batched;
+  objects_batched.push_back(object_id);
+  ASSERT_TRUE(owner_client->ReplyLongPolling(objects_batched));
+  // Since we unsubscribe the object in the subscription callback, there shouldn't be any
+  // long polling request in flight.
+  ASSERT_EQ(owner_client->GetNumberOfInFlightLongPollingRequests(), 0);
+}
+// TODO(sang): Need to add a network failure test once we support network failure
+// properly.
+
+}  // namespace ray
```

Comment:
Can you add a check that all the data structures in pubsub classes are cleaned up (no leaks)? We can assert this at the end of each test.